### ISSUE TITLE
Next.js WVList SSR

### ISSUE THREAD

**Issue by lithdew** [inokawa/virtua#171]
**Describe the bug**
When a WVList is SSR’d on NextJS, the initial HTML markup has all row items rendered as hidden.

InitialItemsCount, and overscan are provided to WVList.

**To Reproduce**
Will provide a reproduction in some time - writing this report on my phone.

**Expected behavior**
All row items should be made visible when SSR rendered.

**Platform:**
 - OS: MacOS
 - Browser: Chrome
 - Version of this package: 0.8.1

**Additional context**
The row items were flexbox elements that have a min-width of zero. Unsure if it is possible for this to be an issue though.

NextJS version is 13.4.19 (the latest).

**Comment by inokawa** [comment]
This lib hide items until items are measured because they can cause visual glitch on initial measuring with current implementation of WVList. And it's impossible to know actual sizes of items in server without layouting on browser (we can know if they are fixed sized items).

So could I ask you the problem in detail caused by hidden items? We may need escape hatch for it.

**Comment by lithdew** [comment]
> This lib hide items until items are measured because they can cause visual glitch on initial measuring with current implementation of WVList. And it's impossible to know actual sizes of items in server without layouting on browser (we can know if they are fixed sized items).
> 
> So could you tell me the problem in detail caused by hidden items? We may need escape hatch for it.

:thinking: So for my use case, I am displaying a post feed for a social network-like website. I would like the list of items pre-rendered via. SSR to improve my site's SEO, and to avoid content layout shift.

Virtual list libraries such as virtuoso seem to handle pre-rendering the list of items just fine, so sizing somehow does happen on server-side (perhaps the initial list of items is SSR'd first, and then sizing happens after?).

**Comment by inokawa** [comment]
react-virtuso will work without hidden style because it has different strategy from virtua.

[As described in this gist](https://gist.github.com/bvaughn/ded0061d712a30c22b0a591cec4aa576), there are 2 major types of strategy of virtual scroller. react-virtuoso seems to be using relative positioning, on the other hand virtua, react-virtualized and react-window using absolute positioning. In absolute positioning without hidden, every items need correct top before initial measuring but not in relative positioning because we can just set position 0 to the spacer if scrolling starts from page top.

```tsx
// absolute
<div style={{ position: 'relative' }}>
  <div style={{ position: 'absolute', top: 0 }}>0</div>
  <div style={{ position: 'absolute', top: 100 }}>1</div>
  <div style={{ position: 'absolute', top: 200 }}>2</div>
  ...
</div>

// relative
<div>
  <div style={{ paddingTop:0, paddingBottom: xxxx }}>
    <div>0</div>
    <div>1</div>
    <div>2</div>
    ...
  </div>
</div>
```

I need some investigation but I'm now thinking is to change strategy of WVList to relative from absolute. Why I adopted absolute strategy is that it was necessary for calculation in scrollToIndex of VList, but WVList doesn't have the method.

**Comment by inokawa** [comment]
As a workaround for fixed sized items, you can override default style of items with `components.Item` prop and provide initial item size in SSR with `initialItemSize` prop:
```tsx
const OverridedItem= forwardRef<HTMLDivElement, CustomItemComponentProps>(
  ({ children, style }, ref) => {
    return (
      <div ref={ref} style={{ ...style, visibility: 'visible' }}>
        {children}
      </div>
    );
  }
);

<VList components={{ Item: OverridedItem }} initialItemCount={10} initialItemSize={100} >{elements}</VList>
```

I think it will only work for fixed sized items and dynamic sized items may cause layout shifts.

**Comment by inokawa** [comment]
Fixed in `0.19.0`.
